testplan/runnable/interactive/reloader.py

```diff
@@ -1,219 +1,530 @@
-                sys.modules['__main__'].__file__)
-            reload_dir = os.path.realpath(os.path.dirname(main_module_file))
+    @property
+    def graph_string(self):
+        """
+        :return: a multi-line string representing the graph from this node
+            downwards.
+        :rtype: ``str``
+        """
+        if self.dependencies:
+            ret_str = "{mod} ->\n{children}".format(
+                mod=self.name,
+                children=strings.indent(
+                    "\n".join(c.graph_string for c in self.dependencies)))
         else:
-            reload_dir = os.path.realpath(pwd())
-        reload_dirs.append(fix_home_prefix(reload_dir))
-
-        # Add extra reload source directories.
-        for mod in extra_deps:
-            module_file = self._module_filepath(mod.__file__)
-            reload_dir = os.path.realpath(os.path.dirname(module_file))
-            reload_dirs.append(fix_home_prefix(reload_dir))
-        return reload_dirs
+            ret_str = self.name
+
+        return ret_str
+
+    @property
+    def name(self):
+        return self.mod.__name__
 
-    def _build_dependencies(self):
+    def reload(self):
+        """Reload this module from file."""
+        self._native_mod = reload_module(self._native_mod)
+
+    def update_suites(self, suite_instances):
         """
-        Builds the dependency tree between the modules.
+        Update suite objects to new classes.
+
+        :param suite_instances: Mapping of module and class name to test suite
+            instances.
+        :type suite_instances: ``Dict[str, Dict[str, List[Any]]``
         """
-        all_files = set()
-        self._dependencies = {}
-        self._files_to_modname = {}
+        if self.name not in suite_instances:
+            return
 
-        for name, mod in sys.modules.items():
-            try:
-                mod_filepath = self._module_filepath(
-                    inspect.getfile(mod))
-                if any(mod_filepath.startswith(directory)
-                       for directory in self.reload_dirs):
-                    all_files.add(mod_filepath)
-                    self._files_to_modname[mod_filepath] = name
-            except:
-                pass
-
-        for filepath in all_files:
-            self._dependencies[filepath] = set()
-            finder = ModuleFinder(path=self.reload_dirs)
-            finder.run_script(filepath)
-
-            for name, mod in finder.modules.items():
-                if name == '__main__':
-                    continue
-                try:
-                    mod_filepath = self._module_filepath(
-                        inspect.getfile(sys.modules[name]))
-                    if any(mod_filepath.startswith(directory)
-                           for directory in self.reload_dirs):
-                        self._dependencies[filepath].add(mod_filepath)
-                except:
-                    pass
-
-        # Manipulating dependencies so that
```

Comment:
Do we no longer need this piece of logic？

Reply:
No, this is the old logic for building dependencies. It is fundamentally incorrect - modulefinder here will find all dependencies of a given script including indirect ones, whereas we only want to build a graph of direct dependencies. Also some modules cannot be run as a script by modulefinder if e.g. it is part of a package and uses relative imports, so we will miss those modules completely from the dependency graph.

By subclassing modulefinder and override the import_hook, we can detect only the direct dependencies and build a graph.